source/common/filter/filter_config_discovery_impl.cc

```diff
@@ -0,0 +1,214 @@
+#include "common/filter/filter_config_discovery_impl.h"
+
+#include "envoy/config/core/v3/extension.pb.validate.h"
+#include "envoy/server/filter_config.h"
+
+#include "common/config/utility.h"
+#include "common/grpc/common.h"
+#include "common/protobuf/utility.h"
+
+#include "absl/strings/str_join.h"
+
+namespace Envoy {
+namespace Filter {
+
+DynamicFilterConfigProviderImpl::DynamicFilterConfigProviderImpl(
+    HttpFilterConfigSubscriptionSharedPtr&& subscription, bool require_terminal,
+    const std::set<std::string>& require_type_urls,
+    Server::Configuration::FactoryContext& factory_context)
+    : subscription_(std::move(subscription)), require_terminal_(require_terminal),
+      require_type_urls_(require_type_urls), tls_(factory_context.threadLocal().allocateSlot()),
+      init_target_("DynamicFilterConfigProviderImpl", [this]() {
+        subscription_->start();
+        init_target_.ready();
+      }) {
+  subscription_->filter_config_providers_.insert(this);
+  tls_->set([](Event::Dispatcher&) -> ThreadLocal::ThreadLocalObjectSharedPtr {
+    return std::make_shared<ThreadLocalConfig>();
+  });
+}
+
+DynamicFilterConfigProviderImpl::~DynamicFilterConfigProviderImpl() {
+  subscription_->filter_config_providers_.erase(this);
+}
+
+const std::string& DynamicFilterConfigProviderImpl::name() { return subscription_->name(); }
+
+absl::optional<Http::FilterFactoryCb> DynamicFilterConfigProviderImpl::config() {
+  return tls_->getTyped<ThreadLocalConfig>().config_;
+}
+
+void DynamicFilterConfigProviderImpl::validateConfig(
```

Comment:
Do we still need this now that we know the type URL constraints?

Reply:
I'm not sure I understand the question. This is enforcing the type URL constraints, and is needed as a callback from xDS to reject responses.

Reply:
As discussed offline, ideally HCM continues to validate termination condition.
/wait

Reply:
Moved termination condition to be conservatively applied across required type URLs.

Reply:
Do we still need this method on the provider interface?

Reply:
Yes, to verify required type URLs. There can be multiple subscriptions on the same resource from multiple HCMs with different required type URLs.

Reply:
OK, but why can't this be done inside of `onConfigUpdate()`?

Reply:
It is called from onConfigUpdate. Providers register with the subscription so that this validation function is called. Are you proposing that the providers modify the subscription instead?

Reply:
Yes, but it could be called from provider `onConfigUpdate()`. Is the idea that you don't want a partial application of the config update on some of the subscriptions prior to validation? If so, I suggest commenting this explicitly.

Reply:
Yeah, I don't want a resource to be accepted by some providers and not others. That creates ambiguity with the state of the resource. I think I mention that a resource is rejected if any provider type URL mismatches, but I can also add code comments.